MatchmakingServer/SignalR/QueueingService.cs

```diff
@@ -0,0 +1,875 @@
+                var joinTriggeredSuccessfully = await _ctx.Clients.Client(player.ConnectionId)
+                    .NotifyJoin(server.GetActualIpAddress(), server.ServerPort, cancellation.Token);
+
+                if (joinTriggeredSuccessfully)
+                {
+                    player.State = PlayerState.Joining;
+                    server.JoiningPlayerCount++;
+
+                    _logger.LogDebug("Player {player} triggered join to {server} successfully.", player, server);
+                }
+                else
+                {
+                    // remove from queue directly when triggering join failed
+                    DequeuePlayer(player, PlayerState.Connected, DequeueReason.JoinFailed, notifyPlayerDequeued: false);
+                }
+            }
+            catch (OperationCanceledException) when (cancellation.IsCancellationRequested)
+            {
+                // timeout
+                _logger.LogDebug("Timed out while waiting for player {player} to join", player);
+
+                DequeuePlayer(player, PlayerState.Connected, DequeueReason.JoinTimeout);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error while notifying player {player} to join", player);
+
+                // TODO: what do here? dequeue player? notify him?
+                DequeuePlayer(player, PlayerState.Connected, DequeueReason.Unknown);
+            }
+        }
+
+        private void ConfirmJoinedPlayers(GameServer server)
+        {
+            _logger.LogDebug("Confirming joined players of {server}", server);
+
+            foreach (var node in server.PlayerQueue.GetNodeSnapshot())
+            {
+                Player player = node.Value;
+
+                if (player.State is PlayerState.Joining &&
+                    server.ActualPlayers.Contains(player.Name))
+                {
+                    if (server.PlayerQueue.TryRemove(node))
+                    {
+                        server.JoiningPlayerCount--;
+                        player.State = PlayerState.Joined;
+                        player.QueuedAt = null;
+
+                        _logger.LogInformation("Confirmed player {player} on server {server}!", player, server);
+                        // yippeeeee
+                    }
+                    else
+                    {
+                        _logger.LogWarning("Player {player} is removed from queue but still in joining state", player);
+                    }
+                }
+            };
+        }
+
+        private async Task UpdateActualPlayersFromWebfront(GameServer server, CancellationToken cancellationToken)
+        {
+            // TODO: fetch actual players
+            _logger.LogDebug("Fetching actual players of server {server}...", server);
+
+            var serverStatusList = await _instanceCache.TryGetWebfrontStatusList(server.InstanceId, cancellationToken);
+            if (serverStatusList.Count == 0)
+            {
```

Comment:
Why is the if statement still there if we don't use it?

Reply:
Just for the comment because empty is also handled in the `FirstOrDefault` below. But yeah thats really stupid :joy: will remove